### ISSUE TITLE
New Feature: Add method for `do()` and/or `group_map()`

### ISSUE THREAD

**Issue by bschneidr** [gergness/srvyr#69]
As mentioned in #68, it would be helpful to have a tidyverse method for complex split-apply-combine operations on grouped data, such as estimating regressions for each of several sub-domains in a survey. 

Currently, `dplyr` has two functions which could be used for this kind of thing in a non-survey context: the `do()` function, and the new experimental `group_map()`, `group_walk()`, and `group_modify()` functions modeled after the `map()` functions from the `purrr` package. It's unclear whether the `do()` functions will eventually be hard-deprecated in favor of the `group_map()` functions. I'd like to implement a `srvyr` version of at least one or the other, but I'm not sure which one would make more sense to start with. I'd appreciate thoughts on this, and hope to implement one approach soon.

Here's what functioning code might look like:
``` r
library(dplyr)
library(survey)
library(srvyr)

data(api)

dclus1 <- apiclus1 %>%
  as_survey_design(ids = dnum,
                   weights = pw,
                   fpc = fpc)

# Use `do()` to run regressions on subsets
dclus1 %>%
   group_by(stype) %>%
   do({
       reg_model <- svyglm(
           design = .,
           formula = api00 ~ meals
       )
       broom::tidy(reg_model)
   })

# Use `group_modify()` to run regressions on subsets
dclus1 %>%
   group_by(stype) %>%
   group_modify(~ {
       reg_model <- svyglm(
           design = .x,
           formula = api00 ~ meals
       )
       broom::tidy(reg_model)
   })
```

**Comment by gergness** [comment]
Available in current development branch. I went with `group_map()`, `group_walk()` matching their dplyr counterparts, and `group_map_dfr()` (instead of `group_modify()`) because I thought "modify" would imply that it returned a srvyr object instead of a `tibble`.

I waited long enough that `do` is back in questioning lifecycle, so I ignored it.


``` r
library(dplyr)
library(survey)
library(srvyr)


data(api)

dclus1 <- apiclus1 %>%
  as_survey_design(ids = dnum,
                   weights = pw,
                   fpc = fpc)

dclus1 %>%
  group_by(stype) %>%
  group_map_dfr(~ {
    reg_model <- svyglm(
      design = .x,
      formula = api00 ~ meals
    )
    broom::tidy(reg_model)
  })
#> # A tibble: 6 x 6
#>   stype term        estimate std.error statistic  p.value
#>   <fct> <chr>          <dbl>     <dbl>     <dbl>    <dbl>
#> 1 E     (Intercept)   833.      13.6       61.5  2.07e-17
#> 2 E     meals          -3.47     0.207    -16.8  3.46e-10
#> 3 H     (Intercept)   763.      37.1       20.6  8.55e- 7
#> 4 H     meals          -3.85     0.610     -6.32 7.33e- 4
#> 5 M     (Intercept)   808.      31.9       25.3  2.14e-10
#> 6 M     meals          -4.10     0.633     -6.47 7.15e- 5
```

<sup>Created on 2020-11-29 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0.9001)</sup>

**Comment by gergness** [comment]
Also note that `summarize` can return multiple row returns now, so some problems may be better solved using that.

 ``` r
dclus1 %>%
  group_by(stype) %>%
  summarize(results = {
    reg_model <- svyglm(
      design = cur_svy(), # `cur_svy()` analagous to `dplyr::cur_data()`
      formula = api00 ~ meals
    )
    broom::tidy(reg_model)
  })
#> # A tibble: 6 x 6
#> # Groups:   stype [3]
#>   stype resultsterm resultsestimate resultsstd.error resultsstatistic
#>   <fct> <chr>                 <dbl>            <dbl>            <dbl>
#> 1 E     (Intercept)          833.             13.6              61.5 
#> 2 E     meals                 -3.47            0.207           -16.8 
#> 3 H     (Intercept)          763.             37.1              20.6 
#> 4 H     meals                 -3.85            0.610            -6.32
#> 5 M     (Intercept)          808.             31.9              25.3 
#> 6 M     meals                 -4.10            0.633            -6.47
#> # … with 1 more variable: resultsp.value <dbl>
```

**Comment by gergness** [comment]
closed via #89
